src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -486,21 +512,76 @@ await _solutionLoadedEvent.WaitAsync()
                             {
                                 if (isAllProjectsNominated)
                                 {
-                                    // if we've got all the nominations then continue with the auto restore
-                                    break;
+                                    if (isBulkRestoreCoordinationEnabled)
+                                    {
+                                        var projectReadyCheckMeasurement = Stopwatch.StartNew();
+                                        if (bulkRestoreCoordinationCheckStartTime == default)
+                                        {
+                                            bulkRestoreCoordinationCheckStartTime = DateTime.UtcNow;
+                                        }
+                                        projectsReadyCheckCount++;
+                                        // If we are about to start restore, we should run through all the projects to ensure there isn't a pending nomination.
+                                        var restoreProjectInfoSources = _solutionManager.Value.GetAllProjectRestoreInfoSources();
+                                        var allProjectsReady = true;
+                                        var bulkCheckTimeout = false;
+                                        for (int i = 0; i < restoreProjectInfoSources.Count && !bulkCheckTimeout; i++)
+                                        {
+                                            var restoreInfoSource = (IVsProjectRestoreInfoSource)restoreProjectInfoSources[i]; 
+                                            if (restoreInfoSource.HasPendingNomination)
+                                            {
+                                                allProjectsReady = false;
+                                                var timeoutTime = (bulkRestoreCoordinationCheckStartTime - DateTime.UtcNow).Value + BulkRestoreCoordinationTimeout;
+                                                var timeoutTask = Task.Delay(timeoutTime, token);
+                                                var whenNominatedTask = restoreInfoSource.WhenNominated(token);
+
+                                                var result = Task.WhenAny(timeoutTask, whenNominatedTask);
+                                                if (result == timeoutTask)
+                                                {
+                                                    bulkCheckTimeout = true;
+                                                }
```

Comment:
May be because if `allProjectsReady` is true the logic exists the loop setting `restoreReason = ImplicitRestoreReason.ProjectsReady;`. Please let me know if I am missing something here.

``` suggestion
                                                if (result == timeoutTask)
                                                {
                                                    bulkCheckTimeout = true;
                                                }
                                                else
                                                {
                                                    allProjectsReady = true;
                                                }                                         
```

Reply:
It's possible that *not* all projects are ready, and a timeout hasn't been reached yet. 
This is the main scenario of concern (nuget not eagerly restoring), so we need to loop back and repeat the logic again.

Reply:
That's the reason why we're keeping track of `projectsReadyCheckCount`.